File: leet-code/15.三数之和.js
/*
 * @lc app=leetcode.cn id=15 lang=javascript
 *
 * [15] 三数之和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  if (nums.length < 3) return [];
  nums.sort((a, b) => a - b);

  let res = [];
  let idx = 0;
  while (idx < nums.length - 2) {
    if (nums[idx] > 0) break

    let left = idx + 1; right = nums.length - 1;

    while (left < right) {
      const sum = nums[idx] + nums[left] + nums[right];
      if (sum === 0) {
        res.push([nums[idx], nums[left], nums[right]])
        left++
        right--

        while (left < right && nums[left] === nums[left - 1]) { left++ }
        while (left < right && nums[right] === nums[right + 1]) { right-- }
      } else if (sum > 0) {
        right--
        while (left < right && nums[right] === nums[right + 1]) { right-- }
      } else {
        left++
        while (left < right && nums[left] === nums[left - 1]) { left++ }
      }
    }

    idx++
    while (idx < nums.length - 2 && nums[idx] === nums[idx - 1]) { idx++ }
  }
  return res
};
// @lc code=end


// @after-stub-for-debug-begin
module.exports = threeSum;
// @after-stub-for-debug-end

File: leet-code-rs/src/solutions/15.三数之和.rs
/*
 * @lc app=leetcode.cn id=15 lang=rust
 *
 * [15] 三数之和
 */

// @lc code=start
impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
        // 对 nums 从小到大排序
        let mut sort_nums = nums.clone();
        sort_nums.sort_unstable();
        println!("{:?}", sort_nums);

        let len = sort_nums.len();
        let mut res = vec![];

        for i in 0..len - 2 {
            let num = sort_nums[i];
            // let (mut j,mut k) = (i+1, len-1);
            let mut j = i + 1;
            let mut k = len - 1;

            // 跳过重复项
            if i > 0 && sort_nums[i - 1] == num {
                continue;
            }

            // 优化
            if num + sort_nums[j] + sort_nums[j + 1] > 0 {
                continue;
            }

            if num + sort_nums[k] + sort_nums[k - 1] < 0 {
                continue;
            }

            while j < k {
                let sum = sort_nums[j] + sort_nums[k] + num;
                if sum == 0 {
                    res.push(vec![sort_nums[j], sort_nums[k], num]);

                    j += 1;
                    // 跳过重复项
                    while j < k && sort_nums[j] == sort_nums[j - 1] {
                        j += 1;
                    }

                    k -= 1;
                    // 跳过重复项
                    while k > j && sort_nums[k] == sort_nums[k + 1] {
                        k -= 1;
                    }
                } else if sum < 0 {
                    j += 1;
                } else {
                    k -= 1;
                }
            }
        }

        return res;
    }
}
// @lc code=end
